### ISSUE TITLE
Reproducible thread error in bumps on Mac OS (Trac #1084)

### ISSUE THREAD

**Issue by smk78** [SasView/sasview#1135]
Reported by user Judith Witte (using Mac OS 10.13.3).

Was fitting happily in an (unspecified) earlier version of !SasView which then just ''"stopped working"'' (wouldn't open anymore). So upgraded to 4.1.2. Data loads in 4.1.2 but on attempting a fit this is returned:

```text
Error(s) Occurred:
> Traceback (most recent call last):
>   File "sas/sasgui/perspectives/fitting/fit_thread.pyc", line 82, in compute
>   File "sas/sasgui/perspectives/fitting/fit_thread.pyc", line 14, in map_apply
>   File "sas/sasgui/perspectives/fitting/fit_thread.pyc", line 11, in map_getattr
>   File "sas/sascalc/fit/BumpsFitting.pyc", line 275, in fit
>   File "sas/sascalc/fit/BumpsFitting.pyc", line 375, in run_bumps
> AttributeError: 'FitDriver' object has no attribute 'fitter'
```

This looks remarkably similar to #1117 which is only classified as minor, however Judith says this is stopping her from doing any fitting. She also comments that ''"The first error occurs after I choose the model (not clicked on Fit yet)"''. If you look carefully at the screenshots you will note two slightly different errors. So am initially marking this as a blocker pending further investigation as it appears a more serious and widespread problem than in 1061.

@pkienzle has commented:

''"There is an error in bumps, but it isn't being properly reported because of an error in the bumps wrapper. I can arrange it so that the next bumps release fixes the wrapper error but I can't address the bumps error without being able to reproduce it."''

Data and screenshots from Judith to assist diagnosis are attached.

Migrated from http://trac.sasview.org/ticket/1084
```json
{
    "status": "assigned",
    "changetime": "2019-03-29T23:45:19",
    "_ts": "2019-03-29 23:45:19.408689+00:00",
    "description": "Reported by user Judith Witte (using Mac OS 10.13.3).\n\nWas fitting happily in an (unspecified) earlier version of !SasView which then just ''\"stopped working\"'' (wouldn't open anymore). So upgraded to 4.1.2. Data loads in 4.1.2 but on attempting a fit this is returned:\n\n{{{\nError(s) Occurred:\n> Traceback (most recent call last):\n>   File \"sas/sasgui/perspectives/fitting/fit_thread.pyc\", line 82, in compute\n>   File \"sas/sasgui/perspectives/fitting/fit_thread.pyc\", line 14, in map_apply\n>   File \"sas/sasgui/perspectives/fitting/fit_thread.pyc\", line 11, in map_getattr\n>   File \"sas/sascalc/fit/BumpsFitting.pyc\", line 275, in fit\n>   File \"sas/sascalc/fit/BumpsFitting.pyc\", line 375, in run_bumps\n> AttributeError: 'FitDriver' object has no attribute 'fitter'\n}}}\n\nThis looks remarkably similar to #1061 which is only classified as minor, however Judith says this is stopping her from doing any fitting. She also comments that ''\"The first error occurs after I choose the model (not clicked on Fit yet)\"''. If you look carefully at the screenshots you will note two slightly different errors. So am initially marking this as a blocker pending further investigation as it appears a more serious and widespread problem than in 1061.\n\n@pkienzle has commented:\n\n''\"There is an error in bumps, but it isn't being properly reported because of an error in the bumps wrapper. I can arrange it so that the next bumps release fixes the wrapper error but I can't address the bumps error without being able to reproduce it.\"''\n\nData and screenshots from Judith to assist diagnosis are attached.",
    "reporter": "smk78",
    "cc": "",
    "resolution": "",
    "workpackage": "SasView Bug Fixing",
    "time": "2018-04-03T09:48:17",
    "component": "SasView",
    "summary": "Reproducible thread error in bumps on Mac OS",
    "priority": "blocker",
    "keywords": "",
    "milestone": "SasView 4.3.0",
    "owner": "wojciech",
    "type": "defect"
}
```

**Comment by smk78** [comment]
Trac update at `2018/04/03 09:48:52`: 
* **smk78** changed attachment from "" to "QI-SM-00039968-STEP1-S4-10-15C-kws3-2017-nicos.DAT"
* **smk78** commented:

Data from Judith

**Comment by smk78** [comment]
Trac update at `2018/04/03 09:49:27`: 
* **smk78** changed attachment from "" to "Bildschirmfoto 2018-04-03 um 09.45.29.png"
* **smk78** commented:

Screenshot 1 from Judith

**Comment by smk78** [comment]
Trac update at `2018/04/03 09:49:57`: 
* **smk78** changed attachment from "" to "Bildschirmfoto 2018-04-03 um 09.46.10.png"
* **smk78** commented:

Screenshot 2 from Judith

**Comment by smk78** [comment]
Trac update at `2018/04/03 09:50:32`: 
* **smk78** changed attachment from "" to "Bildschirmfoto 2018-04-03 um 09.46.47.png"
* **smk78** commented:

Screenshot 3 from Judith

**Comment by smk78** [comment]
Trac update at `2018/04/03 09:51:11`: 
* **smk78** changed attachment from "" to "Bildschirmfoto 2018-04-03 um 09.47.04.png"
* **smk78** commented:

Screenshot 4 from Judith

**Comment by butlerpd** [comment]
Trac update at `2018/04/03 13:33:34`: **butler** changed _comment0 from:

> XCode command line tools not installed and OpenCL turned off.  Options are:
> 
> (1) default to opencl on for Mac (we shut it off because Intel HD6000 drivers were failing for some models; Intel has since fixed their drivers, though they may or may not have been distributed to all macs yet).
> 
> (2) force XCode command line tools to install when SasView first starts.
> 
> (3) ship tinycc for mac
> 

to:

> 1523977011278460

**Comment by pkienzle** [comment]
Trac update at `2018/04/03 13:33:34`: **pkienzle** commented:

XCode command line tools not installed and OpenCL turned off.  Options are:

(1) default to opencl on for Mac (we shut it off because Intel HD6000 drivers were failing for some models; Intel has since fixed their drivers, though they may or may not have been distributed to all macs yet).

(2) force XCode command line tools to install when !SasView first starts.

(3) ship tinycc for mac

**Comment by butlerpd** [comment]
Trac update at `2018/04/17 14:54:24`: 
* **butler** changed _comment0 from:

> As per discussion at !SasView call of April 17, 2018 it is agreed that:
> * OpenCL should be left off by default even though the original problem we think has been fixed.  Because the GPU calculations are sensitive to how they are coded for the card, it is always possible that in a future upgrade to cards or the OpenCL stack that our code could start failing "silently."  Users should use the test function built into !SasView before turning on OpenCL.
> * The right answer is the third option of shipping tinycc for mac.  There is a start at this and it should not be too difficult but probably cannot be done for this release
> 
> Thus to resolve this ticket for release 4.2 we will just improve the error reported so that it is more obvious why it failed and create a new ticket for release 4.3 named "Ship tinycc with mac"

to:

> 1523977059263504


* **butler** commented:

As per discussion at !SasView call of April 17, 2018 it is agreed that:
* OpenCL should be left off by default even though the original problem we think has been fixed.  Because the GPU calculations are sensitive to how they are coded for the card, it is always possible that in a future upgrade to cards or the OpenCL stack that our code could start failing "silently."  Users should use the test function built into !SasView before turning on OpenCL.
* The right answer is the third option of shipping tinycc for mac.  There is a start at this and it should not be too difficult but probably cannot be done for this release

Thus to resolve this ticket for release 4.2 we will just improve the error reported so that it is more obvious why it failed and create a new ticket #1146 for release 4.3 named "Ship tinycc with mac"

**Comment by butlerpd** [comment]
Trac update at `2018/04/17 14:56:25`: 
* **butler** changed owner from "pkienzle" to "wojciech"
* **butler** changed status from "new" to "assigned"

**Comment by butlerpd** [comment]
Trac update at `2018/04/25 15:12:19`: **butler** changed _comment0 from:

> Additional option: 
> 
> (4) precompile models and include them with the application.  Set the error message to give instructions for installing XCode command line tools in case they try creating their own C model or download one from the marketplace.
> 
> This option has the advantage that most users won't see it, or at least not until they are already familiar with SasView.

to:

> 1526345272712052

**Comment by pkienzle** [comment]
Trac update at `2018/04/25 15:12:19`: **pkienzle** commented:

Additional option: 

(4) precompile models and include them with the application.  Set the error message to give instructions for installing XCode command line tools in case they try creating their own C model or download one from the marketplace.

This option has the advantage that most users won't see it, or at least not until they are already familiar with !SasView.

**Comment by butlerpd** [comment]
Trac update at `2018/06/12 08:33:13`: 
* **butler** commented:

For release 4.2 Wojcieck has implemented a more helpful warning to the user of why things are failing. Leaving this ticket open but moving to 4.3 as it is not clear that it has been resolved that we won't try to do something more active (i.e. some of the other options above.) If not we can close the ticket.


* **butler** changed milestone from "SasView 4.2.0" to "SasView 4.3.0"

**Comment by pkienzle** [comment]
Trac update at `2019/03/29 23:45:19`: **pkienzle** commented:

[[Image(ticket:1084:Bildschirmfoto 2018-04-03 um 09.45.29.png)]]
[[Image(ticket:1084:Bildschirmfoto 2018-04-03 um 09.46.10.png)]]
[[Image(ticket:1084:Bildschirmfoto 2018-04-03 um 09.46.47.png)]]
[[Image(ticket:1084:Bildschirmfoto 2018-04-03 um 09.47.04.png)]]

**Comment by butlerpd** [comment]
I am going to close this ticket for now @smk78 on the grounds that this ticket goes way back, that the issue with mac compiler has now been known for a while, that we have documented this in many places to help the user and that there are ongoing discussions about "fixing" in 5.x. At any rate given all this it certainly doesn't seem like something that will be fixed in a final 4.x release...
